### PR TITLE
WIP : egs/aspire : added nnet3 xent, sMBR and chain recipes

### PR DESCRIPTION
**Pull request by vijayaditya**
Has generic scripts to copy and combine lattice directories after adding a prefix or suffix. 
Some of the recipe scripts have not been updated.

I will add results once the experiments are complete.

### CODE REVIEW

File: egs/aspire/s5/local/chain/copy_lat_dir.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  fi
+  cp $src_dir/$f $dest_dir/
+done
+
+nj=$(cat $dest_dir/num_jobs)
+mkdir -p $dest_dir/log
+
+$cmd JOB=1:$nj $dest_dir/log/copy_lat.JOB.log \
+  lattice-copy "ark:gunzip -c $src_dir/lat.JOB.gz |" ark,t:- \|\
+    awk -v prefix="$utt_prefix" -v suffix="$utt_suffix" '{if ((NF == 1) && (prev_nf == 0)) { printf("%s%s%s\n", prefix, $0, suffix) } else { print $0}; prev_nf=NF}' \|\
```
Comment (vijayaditya) [inline egs/aspire/s5/local/chain/copy_lat_dir.sh]: @danpovey , @jtrmal To identify the utterance name lines in the lattices I am checking that the number of fields in the current line is 1 and the number of fields in the previous line is 0. This works for lattices currently copied by `lattice-copy`. Is this a safe assumption to make ? If not, is there any other way to affix strings to utt-ids other than adding these options to  `lattice-copy` ?

Comment (vijayaditya) [inline egs/aspire/s5/local/chain/copy_lat_dir.sh]: @danpovey When copying a lattice directory which files should be definitely be copied other than `lat.*.gz tree cmvn_opts splice_opts num_jobs final.mdl` ?

Comment (vijayaditya) [inline egs/aspire/s5/local/chain/copy_lat_dir.sh]: @tomkocse You could use these scripts when writing the multi-condition multi-en recipe ( #870 ).

Comment (danpovey) [inline egs/aspire/s5/local/chain/copy_lat_dir.sh]: Yes, this assumption about identifying the utterance-ids is safe.  An alternative method is to dump lattices to ark,scp (without gzipping them) and to do operations on the scp file instead.
If you copy the splice_opts you should probably copy delta_opts too (although neither of them is guaranteed to exist).

**Comment by vijayaditya** [comment]
```
nnet2: 
%WER 30.8 | 2120 27213 | 75.3 16.2 8.4 6.2 30.8 78.8 | -0.724 | exp/nnet2_multicondition/nnet_ms_a/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_iterfinal_pp_fg/score_13/penalty_0.0/ctm.filt.filt.sys

nnet3:
%WER 31.0 | 2120 27217 | 74.8 16.1 9.1 5.9 31.0 77.9 | -0.707 | exp/nnet3/tdnn/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iterfinal_pp_fg/score_14/penalty_0.0/ctm.filt.filt.sys
```

**Comment by vijayaditya** [comment]
The BLSTMs show significantly better frame accuracy (see fig below) but the WERs are worse.

```
iteration	TDNN	        BLSTM
====================		
100	        34.2	          43.8
200	        32.8	         40.4
300	        32.3	         38.3
400	        31.7	         36.8
500	        31.6	         35.8
600	        31.3	
700	        31.2	
800	        31.1	
900	          ----
1000	31.1	
Final         31	
``` 

screen shot 2016-07-14 at 11 14 38 pm

**Comment by naxingyu** [comment]
What's the scale of aspire? BLSTM might need more epochs in large scale 
recipes, as in fisher_swbd.


在 2016/7/15 11:17, Vijayaditya Peddinti 写道:

**Comment by vijayaditya** [comment]
These recipes use the same hyper-parameters as fshr_swbd. Only 1/4th of the iterations are complete in BLSTM. However my concern is that even though BLSTM has better train/valid accuracies its WER performance is very poor.

**Comment by vijayaditya** [comment]
The BLSTM training logs several red flags. The average parameter change is around 50 (`--max-param-change 2`) for the first 800/2000 iterations. I also see a huge dip in the objfn around iteration 920. However going through the logs I see no major event (e.g. sudden increase in parameter-change). 

Log dir: /export/b13/vpeddinti/lsr/egs/aspire/s5_olddataaugment/exp/nnet3/lstm_bidirectional/log/

screen shot 2016-07-18 at 1 58 56 am
screen shot 2016-07-18 at 1 59 07 am

**Comment by vijayaditya** [comment]
Nice improvement with chain models in ASpIRE.

```bash 
chain (900/1600 iterations): 
%WER 28.2 | 2120 27211 | 77.4 13.4 9.2 5.6 28.2 75.7 | -0.611 | exp/chain/tdnn_7b/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_frshifted_iter900_pp_fg/score_9/penalty_0.25/ctm.filt.filt.sys

xent (final model ~1100 iterations):
%WER 31.0 | 2120 27217 | 74.8 16.1 9.1 5.9 31.0 77.9 | -0.707 | exp/nnet3/tdnn/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iterfinal_pp_fg/score_14/penalty_0.0/ctm.filt.filt.sys

```

**Comment by vijayaditya** [comment]
@danpovey Ready for review.

```bash 
chain (Final model ~1600 iterations): 
%WER 27.8 | 2120 27217 | 78.2 13.6 8.2 6.0 27.8 75.9 | -0.613 | exp/chain/tdnn_7b/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iterfinal_pp_fg/score_9/penalty_0.0/ctm.filt.filt.sys

xent (final model ~1100 iterations):
%WER 31.0 | 2120 27217 | 74.8 16.1 9.1 5.9 31.0 77.9 | -0.707 | exp/nnet3/tdnn/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iterfinal_pp_fg/score_14/penalty_0.0/ctm.filt.filt.sys

```


Note: I have added the BLSTM script but there are not results available as the experiments were paused for Babel evals.

File: egs/aspire/s5/run.sh [hunk 2]
```diff
@@ -156,6 +158,7 @@ steps/train_sat.sh  --cmd "$train_cmd" \
 
 # build silprob lang directory
 local/build_silprob.sh
+exit;
```
Comment (danpovey) [inline egs/aspire/s5/run.sh]: was this supposed to be temporary?

File: egs/wsj/s5/utils/combine_lat_dirs.sh [hunk 3]
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+# Copyright 2016  Xiaohui Zhang  Apache 2.0.
+#                 Vijay Peddinti
+
+# This script operates on lattice directories, such as exp/tri4a_lats
+
```
Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: please add a line or two saying what it does and what its output[s] are.

Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: Also, the general rule is that any script that takes the 'cmd' option should be in steps/.

File: egs/wsj/s5/utils/combine_lat_dirs.sh [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    ark,scp:$temp_dir/lat.$src_id.ark,$temp_dir/lat.$src_id.scp || exit 1;
+done
+cat $temp_dir/lat.*.scp | sort -m > $temp_dir/lat.scp || exit 1;
+
+echo "$0: splitting data to get reference utt2spk for individual lat.JOB.gz files."
+utils/split_data.sh $data $num_jobs || exit 1;
+echo $num_jobs > $dest/num_jobs  || exit 1
+
+echo "$0: splitting the lattices to appropriate chunks according to the reference utt2spk files."
+for i in `seq 1 $num_jobs`; do
```
Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: Here,  you should be able to use filter_scps.pl, which is more efficient.

File: egs/wsj/s5/utils/combine_lat_dirs.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+echo "$0: dumping lattices in each source directory as single archive and index."
+for dir in $*; do
+  src_id=$((src_id + 1))
+  cur_num_jobs=$(cat $dir/num_jobs) || exit 1;
+  all_ids=$(seq -s, $cur_num_jobs)
+  $cmd $dir/log/copy_lattices.log \
+    lattice-copy "ark:gunzip -c $dir/lat.{$all_ids}.gz|" \
+    ark,scp:$temp_dir/lat.$src_id.ark,$temp_dir/lat.$src_id.scp || exit 1;
+done
+cat $temp_dir/lat.*.scp | sort -m > $temp_dir/lat.scp || exit 1;
```
Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: sort -m doesn't work like this, you have to give it the individual inputs, like
sort -m $temp_dir/lat.*.scp

File: egs/wsj/s5/utils/combine_lat_dirs.sh (hunk 4 above)
Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: also, there is no need to create the utt_subset files.  filter_scp.pl and filter_scps.pl automatically take the first field of the filter file.

File: egs/aspire/s5/run.sh (hunk 2 above)
Comment (vijayaditya) [inline egs/aspire/s5/run.sh]: Yes, thanks for pointing out.

File: egs/wsj/s5/utils/combine_lat_dirs.sh (hunk 4 above)
Comment (vijayaditya) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: Sorry I based this script on `utils/combine_ali_dirs.sh` just replacing alignment commands with lattice commands. I will add the suggested changes to both these scripts.

Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: @xiaohui-zhang, please remember these comments on the script
combine_ali_dirs.sh for the future, although Vijay will fix it.

Where do you use that script?  I can't find anything that uses it.

Comment (xiaohui-zhang) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: Thanks. Hmm it's being used in babel multi-condition experiments . (will commit it in several days)

Comment (danpovey) [inline egs/wsj/s5/utils/combine_lat_dirs.sh]: Vijay, I notice that in egs/aspire/s5/local/multi_condition/ you have your
own combine_ali_dirs.sh script, and it's very hard to understand with a big
inline python section.  I think it would be better to have a single
definitive script to do this, located in steps/.

**Comment by vijayaditya** [comment]
I am waiting for #706 to be merged. After this I will be updating the data preparation code and alignment generation code.

**Comment by vijayaditya** [comment]
Deleted `local/multi_condition/combine_ali_dirs.sh`, modified `steps/combine_ali_dirs.sh` and deleted `steps/combine_lat_dirs.sh` as I turned to be unnecessary for the recipe.

File: egs/wsj/s5/utils/combine_ali_dirs.sh [hunk 6]
```diff
@@ -1,99 +0,0 @@
...
-# End configuration section.
-echo "$0 $@"  # Print the command line for logging
-
-if [ -f path.sh ]; then . ./path.sh; fi
-. parse_options.sh || exit 1;
-
-if [[ $# -lt 3 ]]; then
-  echo "Usage: $0 [options] <data> <dest-ali-dir> <src-ali-dir1> <src-ali-dir2> ..."
-  echo "e.g.: $0 --num-jobs 32 data/train exp/tri3_ali_combined exp/tri3_ali_1 exp_tri3_ali_2"
-  echo "Options:"
```
Comment (danpovey) [inline egs/wsj/s5/utils/combine_ali_dirs.sh]: I think you forgot to 'git add' this in its location in steps.

Comment (vijayaditya) [inline egs/wsj/s5/utils/combine_ali_dirs.sh]: Sorry about that, the file is available now.

File: egs/wsj/s5/steps/combine_ali_dirs.sh [hunk 7]
```diff
@@ -61,6 +62,7 @@ done
 
 src_id=0
 temp_dir=$dest/temp
+rm -rf $temp_dir
```
Comment (danpovey) [inline egs/wsj/s5/steps/combine_ali_dirs.sh]: No -f flag on rm, please, better to do
[ -d $temp_dir ] && rm -r $temp_dir
I just don't like using the -f flag on 'rm' as a general policy.

File: egs/wsj/s5/steps/combine_ali_dirs.sh [hunk 8]
```diff
@@ -71,23 +73,25 @@ for dir in $*; do
     copy-int-vector "ark:gunzip -c $dir/ali.{$all_ids}.gz|" \
     ark,scp:$temp_dir/ali.$src_id.ark,$temp_dir/ali.$src_id.scp || exit 1;
 done
-cat $temp_dir/ali.*.scp | sort -m > $temp_dir/ali.scp || exit 1;
+sort -m $temp_dir/ali.*.scp > $temp_dir/ali.scp || exit 1;
 
 echo "$0: splitting data to get reference utt2spk for individual ali.JOB.gz files."
 utils/split_data.sh $data $num_jobs || exit 1;
-echo $num_jobs > $dest/num_jobs  || exit 1
```
Comment (danpovey) [inline egs/wsj/s5/steps/combine_ali_dirs.sh]: There may be situations where we need this 'num_jobs'... better to keep that.

File: egs/wsj/s5/steps/combine_ali_dirs.sh [hunk 9]
```diff
@@ -71,23 +73,25 @@ for dir in $*; do
...
+
 for i in `seq 1 $num_jobs`; do
-  awk '{print $1}' $data/split$num_jobs/$i/utt2spk | sort > $temp_dir/utt_subset.$i
-  utils/filter_scp.pl $temp_dir/utt_subset.$i $temp_dir/ali.scp | \
-    copy-int-vector scp:- "ark:|gzip -c >$dest/ali.$i.gz" || exit 1;
-done   
+    copy-int-vector scp:$temp_dir/ali.${i}.scp "ark:|gzip -c >$dest/ali.$i.gz" || exit 1;
+done
+
+echo $num_jobs > $dest/num_jobs  || exit 1
```
Comment (danpovey) [inline egs/wsj/s5/steps/combine_ali_dirs.sh]: .. it's OK, I see you are doing it here.

File: egs/wsj/s5/steps/nnet3/decode.sh [hunk 10]
```diff
@@ -80,7 +80,7 @@ thread_string=
 [ $num_threads -gt 1 ] && thread_string="-parallel --num-threads=$num_threads"
 
 mkdir -p $dir/log
-[[ -d $sdata && $data/feats.scp -ot $sdata ]] || split_data.sh $data $nj || exit 1;
+[[ -d $sdata && $data/feats.scp -ot $sdata ]] || split_data.sh --per-utt $data $nj || exit 1;
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/decode.sh]: Was the addition of this --per-utt flag intended to be temporary?
Until issue #778 is addressed, this is not safe to commit as it would interfere with jobs that do require splitting per-speaker.

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/decode.sh]: I intended it to be permanent as the ASpIRE decodes take several hours when
done per speaker. I will remove it from the script as soon as I get access
to a computer.

Vijay

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/decode.sh]: We've been waiting too long for the per-utt splitting to be working.
I'll change the data-splitting script right now and check it in within the
hour.
After that you can set something like
sdata=$dir/split${nj}utt
instead of
sdata=$dir/split${nj}
and it will work.
Dan

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/decode.sh]: OK thanks.

Vijay

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/decode.sh]: ok, committed that change.

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/decode.sh]: You now have to change `sdata=$data/split$nj;` to `sdata=$data/split${nj}utt;`

File: egs/wsj/s5/steps/nnet3/decode.sh [hunk 11]
```diff
@@ -80,7 +80,7 @@ thread_string=
 [ $num_threads -gt 1 ] && thread_string="-parallel --num-threads=$num_threads"
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/decode.sh]: vijay, I think we can merge this once you make this small change... a few lines above here, add 'utt' at the end of 'sdata'.

**Comment by vijayaditya** [comment]
@danpovey All the requested changes have been done. Sorry for the delay, I was unable to test out the final changes due to IO latency issues on CLSP cluster till now.

The recipe in the current form is not going to be able to reproduce the results due to the bug in `wav-reverberate` command. This bug has been corrected in #706 , which also does other major modifications to the data reverberation scripts. So I am not going to address this issue in this PR.
